docs/site/Interceptors.md

```diff
@@ -0,0 +1,529 @@
+permalink: /doc/en/lb4/Interceptors.html
+---
+
+## Overview
+
+Interceptors are reusable functions to provide aspect-oriented logic around
+method invocations. There are many use cases for interceptors, such as:
+
+- Add extra logic before / after method invocation, for example, logging or
+  measuring method invocations.
+- Validate/transform arguments
+- Validate/transform return values
+- Catch/transform errors, for example, normalize error objects
+- Override the method invocation, for example, return from cache
+
+## Basic use
+
+### Interceptors on controllers
+
+Interceptors are supported for public controller methods (including both static
+and prototype) and handler functions for REST routes.
+
+Controller methods decorated with `@intercept` are invoked with applied
+interceptors for corresponding routes upon API requests.
+
+```ts
+import {intercept} from '@loopback/context';
+
+@intercept(log)
+export class OrderController {
+  @intercept('caching-interceptor')
+  async listOrders(userId: string) {
+    // ...
+  }
+}
+```
+
+It's also possible to configure global interceptors that are invoked before
+method level interceptors. For example, the following code registers a global
+`caching-interceptor` for all methods.
+
+```ts
+app
+  .bind('caching-interceptor')
+  .toProvider(CachingInterceptorProvider)
+  .apply(asInterceptor);
+```
+
+Global interceptors are also executed for route handler functions without a
+controller class. See an example in
+[Route Handler](Routes.md#using-partial-openapi-spec-fragments).
+
+### Create a proxy to apply interceptors
+
+A
+[proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
+can be created from the target class or object to apply interceptors. This is
+useful for the case that a controller declares dependencies of repositories or
+services and would like to allow repository or service methods to be
+intercepted.
+
+```ts
+import {createProxyWithInterceptors} from '@loopback/context';
+
+const proxy = createProxyWithInterceptors(
+  controllerInstanceWithClassInterceptors,
+  ctx,
+);
+const msg = await proxy.greet('John');
+```
+
+There is also an `asProxyWithInterceptors` option for binding resolution or
+dependency injection to return a proxy to apply interceptors as methods are
+invoked.
+
+```ts
+class DummyController {
+  constructor(
+    @inject('my-controller', {asProxyWithInterceptors: true})
+    public readonly myController: MyControllerWithClassLevelInterceptors,
+  ) {}
+}
+ctx.bind('my-controller').toClass(MyControllerWithClassLevelInterceptors);
+ctx.bind('dummy-controller').toClass(DummyController);
+const dummyController = await ctx.get<DummyController>('dummy-controller');
+const msg = await dummyController.myController.greet('John');
+);
+```
+
+Or:
+
+```ts
+const proxy = await ctx.get<MyControllerWithClassLevelInterceptors>(
+  'my-controller',
+  {asProxyWithInterceptors: true},
+const msg = await proxy.greet('John');
```

Comment:
Syntax error. Should be:

```diff
 const proxy = await ctx.get<MyControllerWithClassLevelInterceptors>(
   'my-controller',
   {asProxyWithInterceptors: true},
+ );
 const msg = await proxy.greet('John');
```

I spotted the error because the markdown source rendered by GitHub in the PR diff is using incorrect colors from the next line onwards: the text is in blue instead of in black. 

When you fix this error then please check the rest of this file to ensure there are no other typos in the code snippets.